src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferAsSpanOverSubstring.cs

```diff
@@ -0,0 +1,294 @@
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(Resx.PreferAsSpanOverSubstringDescription), Resx.ResourceManager, typeof(Resx));
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!RequiredSymbols.TryGetSymbols(context.Compilation, out RequiredSymbols symbols))
+                return;
+
+            context.RegisterOperationBlockStartAction(OnOperationBlockStart);
+
+            void OnOperationBlockStart(OperationBlockStartAnalysisContext context)
+            {
+                var invocations = PooledConcurrentSet<IInvocationOperation>.GetInstance();
+
+                context.RegisterOperationAction(context =>
+                {
+                    var argument = (IArgumentOperation)context.Operation;
+                    if (symbols.IsAnySubstringInvocation(argument.Value.WalkDownConversion(c => c.IsImplicit)) && argument.Parent is IInvocationOperation invocation)
+                    {
+                        invocations.Add(invocation);
+                    }
+                }, OperationKind.Argument);
+
+                context.RegisterOperationBlockEndAction(context =>
+                {
+                    foreach (var invocation in invocations)
+                    {
+                        //  We search for an overload of the invoked member whose signature matches the signature of
+                        //  the invoked member, except with ReadOnlySpan<char> substituted in for some of the 
+                        //  arguments that are Substring invocations.
+                        if (!GetBestSpanBasedOverloads(symbols, invocation, context.CancellationToken).IsEmpty)
+                        {
+                            Diagnostic diagnostic = invocation.CreateDiagnostic(Rule);
+                            context.ReportDiagnostic(diagnostic);
+                        }
+                    }
+                    invocations.Free(context.CancellationToken);
```

Comment:
Nit: add newline between closing brace and next statement
```suggestion

                    invocations.Free(context.CancellationToken);
```

Reply:
Fixed